packages/salesforcedx-vscode-apex/src/views/testOutline.ts

```diff
@@ -0,0 +1,666 @@
+      // Starting Out
+      this.head = new ApexTestGroup('ApexTests', null, 1);
+    }
+    // Check if lsp worked, else parse files manually
+    if (this.apexTestInfo) {
+      // Do it the easy way
+      this.apexTestInfo.forEach(test => {
+        let apexGroup = this.apexTestMap.get(test.parent) as ApexTestGroup;
+        if (!apexGroup) {
+          apexGroup = new ApexTestGroup(test.parent, null, 1);
+          this.apexTestMap.set(test.parent, apexGroup);
+        }
+        const testUri = vscode.Uri.file(test.file.substring(7));
+        const apexTest = new ApexTest(test.methodName, testUri, test.line);
+        apexTest.name = apexGroup.label + '.' + apexTest.label;
+        this.apexTestMap.set(apexTest.name, apexTest);
+        apexGroup.children.push(apexTest);
+        if (this.head && !(this.head.children.indexOf(apexGroup) >= 0)) {
+          this.head.children.push(apexGroup);
+        }
+        ApexTestOutlineProvider.testStrings.add(apexGroup.name);
+      });
+    } else {
+      this.apexClasses.forEach(apexClass => {
+        const fileContent = fs.readFileSync(apexClass.fsPath).toString();
+        if (fileContent && fileContent.toLowerCase().includes('@istest')) {
+          const testName = ospath
+            .basename(apexClass.toString())
+            .replace('.cls', '');
+          const newApexTestGroup = new ApexTestGroup(testName, apexClass, 1);
+          this.apexTestMap.set(testName, newApexTestGroup);
+          this.addTests(fileContent, newApexTestGroup);
+          ApexTestOutlineProvider.testStrings.add(testName);
+          if (this.head) {
+            this.head.children.push(newApexTestGroup);
+          }
+        }
+      });
+    }
+    return this.head;
+  }
+
+  private addTests(fileContent: string, apexTestGroup: ApexTestGroup): void {
+    // Parse through file and find apex tests
+    const regEx = /(@[iI][sS][tT][eE][sS][tT])+/g;
+    let match = regEx.exec(fileContent);
+    while (match) {
+      const ind = match.index;
+      // Ignore class heading
+      if (ind !== 0) {
+        // A test
+        const restOfFile = fileContent.slice(match.index);
+        const header = restOfFile.slice(0, restOfFile.indexOf('{'));
+        const headerParts = header.split(' ');
+
+        // Name is last part of the header, not always the last item though, because could be an empty string ''
+        let name = '';
+        while (name === '') {
+          const headerPart = headerParts.pop();
+          if (headerPart && headerPart.includes(')')) {
+            name = headerPart;
+          }
+        }
+        name = name.replace('()', ''); // Get rid
+        const apexTest = new ApexTest(name, apexTestGroup.file, ind);
+        apexTest.name = apexTestGroup.label + '.' + name;
+        this.apexTestMap.set(apexTest.name, apexTest);
+        apexTestGroup.children.push(apexTest);
+      }
+      match = regEx.exec(fileContent);
+    }
+  }
+
+  private deleteFolderRecursive(path: string) {
```

Comment:
What is this for? What are you deleting?

Reply:
I output results to a temporary output directory, and then I go and delete it. This is so that we can still output the human output to the console. @cwallsfdc mentioned that this would be a good option to use

Reply:
Is there no utility that we can use to delete this short of writing our own recursive delete? Maybe you can search for `shelljs` and see how we use it there.

Reply:
Also, if you are using the OS temp folders, aren't those deleted automatically for us?

Reply:
shell.rm('-rf', path) is not working. I could use the /tmp OS temp folder for Mac and Linux. For windows, the temp folder is stored in the TEMP environment variable. Not sure if that is a good approach either

Reply:
All right, let's tackle the issue holistically here since I might be missing something.

Intent
===
Purpose (please confirm/refute): we need a way to store temporary results so that we can still output human results (since JSON is not human readable results, I assume). 

Questions:
1. How bad is the JSON results from a human-readable perspective?
1. How come we didn't choose to use the TAP format from the test runner? That one seems to be able to stream results through and is also pretty human-readable.


Implementation
===

If we do need to store the results somewhere temporarily, could we use some of the features of the OS to simplify our implementation? Though, some posts like [this](https://superuser.com/questions/296824/when-is-a-windows-users-temp-directory-cleaned-out) seems to suggest that Windows doesn't clean out the temp folders automatically.

1. Does https://nodejs.org/docs/latest-v7.x/api/os.html#os_os_tmpdir not work? Note, I am using Node 7's doc because that is the version of Node that VS Code ships with as of August 2018.
1. "shell.rm('-rf', path) is not working" <-- what is the error that you are seeing?

Reply:
Intent:
1. The JSON results are definitely not very readable, and especially if I want to later allow any invocation of these tests to update the test view (such as the code lens button).
2. I followed the example of the other run test examples in the extension.

Implementation:

1. Yes, that works. I will have to do some tests on Windows to see if that folder gets cleaned automatically.

2. There is no elaboration on the error, even if I turn on the developer tools in the extension host, it just does not remove the folder.

Reply:
"I want to later allow any invocation of these tests to update the test view (such as the code lens button)." <-- this is a good point. Does the code lens button fire off any events now that the test view can list for?

Reply:
Well, it uses the forceApexTestRunCodeActionExecutor, so I was going to utilize that